[gohsl99] iP

gohsl99 opened the pull request:
# Blitz
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Blitz frees your mind of having to remember things you need to do. It's,
- text-based
* easy to learn
+ ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Duke.java: Rename to Blitz.java and implement initial skeletal version
README.md, docs/README.md, text-ui-test/runtest.bat: Replace Duke with Blitz

Commit:
Blitz.java: Add echo functionality

Commit:
Blitz.java: Add add and list functionality

Commit:
Blitz.java: Add mark task as done and not done functionality, and improve output text
Task.java: Add new class to help achieve functionality

Commit:
Add new classes: Deadline.java, Event.java, Todo.java
Add support for more types of task

Commit:
Add automated text UI testing

Commit:
Blitz.java: Add handle errors functionality

Commit:
Blitz.java: Add Enum usage
Blitz*Exception.java: Improve Level-5 with exception classes

Commit:
Blitz.java: Add delete functionality

Commit:
Add basic save task functionality
Add addition custom exception class

Commit:
Update delimiter to save to file
Add functionality to handle situation of corrupted data file

Commit:
Merge branch 'branch-Level-7'

Commit:
Add date time functionality for deadline command

Commit:
Add date time functionality for event command

Commit:
Merge branch 'branch-Level-8'

Commit:
Add more OOP by separating functionality using different classes (in progress)

Commit:
Add more OOP by separating functionality using different classes (completed)

Commit:
Modify gitignore file to ignore .class file and Blitz.txt

Commit:
Merge branch 'branch-more-oop'

Commit:
Organize classes into packages
Fix minor bug (bye not working)

Commit:
Merge branch 'add-gradle-support'

Commit:
Integrate Gradle into project (build and run)

Commit:
Add Junit tests for Blitz package (src/test/java/Blitz/*.java)
Modify classes for Junit tests

Commit:
Modify build.gradle for generating jar file

Commit:
Modify code to follow coding standards

Commit:
Modify code to follow coding standards

Commit:
Modify code to follow coding standards

Commit:
Add javadoc for package blitz

Commit:
Add javadoc for package command

Commit:
Add javadoc for package exception

Commit:
Add javadoc for package task

Commit:
Add find task functionality

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Blitz/Blitz.java
#	src/main/java/Blitz/Parser.java
#	src/test/java/Blitz/UiTest.java

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Blitz/Instruction.java
#	src/main/java/Blitz/Parser.java
#	src/main/java/Blitz/Storage.java
#	src/main/java/Blitz/TaskList.java
#	src/main/java/Blitz/Ui.java
#	src/main/java/Command/Command.java
#	src/main/java/Command/CommandBye.java
#	src/main/java/Command/CommandDeadline.java
#	src/main/java/Command/CommandDelete.java
#	src/main/java/Command/CommandEvent.java
#	src/main/java/Command/CommandList.java
#	src/main/java/Command/CommandMark.java
#	src/main/java/Command/CommandTodo.java
#	src/main/java/Command/CommandUnmark.java
#	src/main/java/Task/Deadline.java
#	src/main/java/Task/Event.java
#	src/main/java/Task/Task.java
#	src/test/java/Blitz/UiTest.java

Commit:
Add javadoc for CommandFind.java
Add gradle test for CommandFind

Commit:
Add checkstyle using gradle
Add JavaDoc for all the class

Domokunx commented on src/main/java/Blitz/Blitz.java:
```diff
@@ -0,0 +1,73 @@
+            "Hello! I'm Blitz.",
+            "What can I do for you?"
+        };
+        String[] end = {"Bye. Hope to see you again soon!"};
```
Could you explain why this is a String[]?
Additionally, consider using plurals for Collections and Lists.

Domokunx commented on src/main/java/Blitz/Blitz.java:
```diff
@@ -0,0 +1,73 @@
+            try {
+                Command c = Parser.parse(inp);
+
+                if (c.isExit()) {
```
Could consider moving c.isExit into the predicate of the while loop at Line 44 :)

Domokunx commented on src/main/java/Blitz/Blitz.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    /**
+     * Entry point of the application.
```
```suggestion
     * Runs/Starts the application
```

Domokunx commented on src/main/java/Blitz/Instruction.java:
```diff
@@ -0,0 +1,34 @@
+     * @param command Command String to be checked.
+     * @return True if the command exists, false otherwise.
+     */
+    public static boolean isCommandExist(String command) {
```
```suggestion
    public static boolean isValidCommand(String command) {
```
just a nitpick :p

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+import command.CommandList;
+import command.CommandMark;
+import command.CommandTodo;
+import command.CommandUnmark;
```
Consider adding a new line after this to separate different packages.

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+import exception.BlitzNoParameterException;
+
+/**
+ * Responsibles for parsing command String and converting it into corresponding Command Object.
```
```suggestion
 * Represents the object that can parse command String and convert it into corresponding Command Object.
```

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+     */
+    public static Command parse(String command) throws BlitzException {
+        if (isRegexMatched("^list\\s*$", command)) {
+            return new CommandList("list");
```
Do you think that ListCommand is a more natural ordering of the words? Just a thought

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+        } else if (isRegexMatched("^bye\\s*$", command)) {
+            return new CommandBye("bye");
+        } else {
+            String[] cont = command.split(" ", 2);
```
Collections should have plural naming to be more clear. Maybe "inputs" could work better?

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                throw new BlitzNoParameterException();
+            }
+
+            switch (inst) {
```
Is there a reason for why "list" and "bye" are not part of this switch case?

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+
+            switch (inst) {
+            case "mark":
+                String[] markParam = cont[1].split(" ");
```
```suggestion
                String[] markParams = cont[1].split(" ");
```

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+
+                return new CommandMark(command, markParam[0]);
+            case "unmark":
+                String[] unmarkParam = cont[1].split(" ");
```
```suggestion
                String[] unmarkParams = cont[1].split(" ");
```

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                return new CommandTodo(command, cont[1]);
+            case "deadline":
+                if (!isRegexMatched(
+                        "^.+ \\/by (19|20)\\d\\d-(0[1-9]|1[0-2])-(0[1-9]|[12]\\d|3[01])"
```
A comment to explain this Regex would be helpful :)

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                        "^.+ \\/by (19|20)\\d\\d-(0[1-9]|1[0-2])-(0[1-9]|[12]\\d|3[01])"
+                                + " (0[0-9]|1[0-9]|2[0-3])[0-5][0-9]$",
+                        command
+                )) {
```
Move the 2 closing parenthesis to the previous line might be better

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                    );
+                }
+
+                String[] param1 = cont[1].split(" /from ");
```
param1 and param2 are too similar, not distinct enough to understand what they actually contain

Maybe fromParams and toParams could work?

Domokunx commented on src/main/java/Blitz/Parser.java:
```diff
@@ -0,0 +1,128 @@
+
+                return new CommandEvent(command, eventParams);
+            case "delete":
+                String[] deleteParam = cont[1].split(" ");
```
```suggestion
                String[] deleteParams = cont[1].split(" ");
```

Domokunx commented on src/main/java/Blitz/Storage.java:
```diff
@@ -0,0 +1,98 @@
+     * @param list TaskList that contains all the Task objects to be written to the file.
+     * @throws BlitzException If I/O error occurs.
+     */
+    public void writeAllToFile(TaskList list) throws BlitzException {
```
throwing BlitzException might be too generic? 
Better to be able to narrow down if its only 1 to 2 exceptions.

Domokunx commented on src/main/java/Command/CommandDelete.java:
```diff
@@ -0,0 +1,61 @@
+            Task task = list.deleteTask(ind);
+            storage.writeAllToFile(list);
+
+            String[] toPrint = {
```
```suggestion
            String[] sentences = {
```

Commit:
Integrate use of GUI in Blitz application

Commit:
Enhance UI

Commit:
Add Varargs usage for deadline and event command

Commit:
Merge branch 'branch-Level-10'

Domokunx commented on src/main/java/Command/CommandFind.java:
```diff
@@ -0,0 +1,63 @@
+            throw new BlitzEmptyTaskListException();
+        }
+
+        String toPrint = "Here are the matching tasks in your list:\n";
```
Consider moving these print functionalities into your Ui class for to make it more easily modified in the future

Domokunx commented on src/main/java/Blitz/Ui.java:
```diff
@@ -0,0 +1,21 @@
+                + "  [" + type + "][ ] " + task + "\n"
+                + "Now you have " + size + " tasks in the list.";
+    }
+}
```
Feel like this class is underused and could afford to contain more of the console output printing functionalities

Domokunx commented on src/main/java/Command/CommandList.java:
```diff
@@ -0,0 +1,48 @@
+import blitz.Storage;
+import blitz.TaskList;
+import blitz.Ui;
+import exception.BlitzEmptyTaskListException;
```
Could consider separating the import statements by package

Domokunx commented on src/main/java/Exception/BlitzException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represents a generic exception in the Blitz application.
+ */
+public class BlitzException extends Exception {
```
All of the exceptions under BlitzException feel like RuntimeExceptions rather than checked exceptions.

Consider extending RuntimeException instead?

Domokunx reviewed (commented):
Overall clean and easy to read! 
Good job on the JavaDocs as well 😄 

Consider refactoring more into your Ui class, and maybe changing your BlitzException into a RuntimeException

Commit:
Improve coding standard

Commit:
Improve code quality

Change the file names related to Command *Command to make it more
understandable.

Use new variables in Equals method in Deadline, Event and Todo to make
the return statement more understandable.

Introduce function wrapper for isRegexMatched method in Parser to make
the condition more understandable.

Commit:
Replace implementation of find command to use stream

Commit:
Replace contains method in Instruction to use stream

Commit:
Add assertions to relevant code

Commit:
Add code quality

Assign a variable name for parameters used for created Task. Doing so
will make the code more readable.

gohsl99 committed:
Merge pull request #1 from gohsl99/branch-A-Assertions

Add assertions to relevant code

gohsl99 committed:
Merge branch 'master' into branch-A-Streams

gohsl99 committed:
Merge pull request #2 from gohsl99/branch-A-Streams

Add streams functionality

Commit:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/Blitz/Blitz.java
#	src/main/java/Blitz/Parser.java
#	src/main/java/Command/FindCommand.java

Commit:
Fix regression error

There are several errors resulted from merging 3 branches. They are not
major errors, just some careless mistakes.

JUnit tests are able to pass now.

Will do further in-depth checking of code to continue to improve.

Commit:
Improve code quality

Assign variable name to parameter values so that they are more readable.

Commit:
Add CI

Commit:
Add detection of duplicates of task

User should not be able to add duplicated task into the list (regardless
the task in the list is marked done or undone).

Application now is able to detect such duplicates. Exception will be
thrown if user trying to add duplicated task.

Another comparing method is added to Task objects to not compare status
unlike the overridden equals method.

gohsl99 committed:
Merge pull request #4 from gohsl99/branch-BCD-Extension

Add detection of duplicates of task

Commit:
Fix CI not working

Commit:
Fix typo

Commit:
Improve code quality

Commit:
Modify UI string format

gohsl99 committed:
Rename BlitzIOException.java to BlitzIoException.java

gohsl99 committed:
Change class and package name

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

gohsl99 committed:
Update README.md

Commit:
Fix code quality issue

Commit:
Add more testing for UI

Commit:
Fix minor bug

Commit:
Add Ui png and title

Commit:
Fix checkstyle custom import

Commit:
Fix Javadoc

gohsl99 committed:
Update README.md

gohsl99 committed:
Change package name

Commit:
Fix some bugs

Commit:
Fix minor bug and JUnit test

Commit:
Improve code quality

Commit:
Improve code quality
